### PR TITLE
MultiLabelDotProductLoss for IntentTED (new)

### PR DESCRIPTION
**Pull request by JEM-Mosig**
**Proposed changes**:
- Make `DotProductLoss` an abstract meta class for `SingleLabelDotProductLoss` (the old one, as used by TED) and `MultiLabelDotProductLoss` (new class, used for IntentTED)
- Implement clean code for MultiLabelDotProductLoss and add tests
- Closes https://github.com/RasaHQ/rasa/issues/8438

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/utils/tensorflow/layers.py [hunk 1]
```diff
@@ -1085,3 +1140,314 @@ def call(
...
+        """Returns batch of random candidate indices
+        
+        Args:
+            batch_size: Size of zero'th dimension
+            total_candidates: Size of first dimension
+
+        Returns:
+            tf.int32 tensor of shape (target_size, total_candidates) with values 
+            in [0, self.num_neg).
+        """
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/layers.py:1343]: Do we need this as an extra function? Can't we directly call `layer_utils.random_indices` just as how `_get_negs` of `SingleLabelDotProductLoss` class does?

File: rasa/utils/tensorflow/layers.py [hunk 2]
```diff
@@ -1085,3 +1140,314 @@ def call(
...
+        x: tf.Tensor,  # (batch_size, ...)
+        candidate_ids: tf.Tensor,  # (batch_size, num_candidates)
+    ) -> tf.Tensor:
+        """Gathers candidate values according to IDs.
+        
+        Args:
+            x: Any tensor with at least one dimension
+            candidate_ids: Indicator for which candidates to gather
+
+        Returns:
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/layers.py:1357]: This one too. Can't we reuse `_sample_idxs` function from `SingleLabelDotProductLoss` by moving it to the base class and reusing it? We might need to change which axis is expanded / squeeze to get a general implementation but the idea of both the functions is the same, i.e. - Given a tensor and some ids, gather the values at those ids from the tensor.

File: rasa/utils/tensorflow/layers.py [hunk 3]
```diff
@@ -1085,3 +1140,314 @@ def call(
...
+
+        if mask is not None:
+            loss *= mask
+
+        if len(loss.shape) == 2:
+            # Average over the sequence
+            if mask is not None:
+                loss = tf.reduce_sum(loss, axis=-1) / tf.reduce_sum(mask, axis=-1)
+            else:
+                loss = tf.reduce_mean(loss, axis=-1)
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/layers.py:1424]: This part is also overlapping with `_loss_cross_entropy` and hence can be reused by moving it in a separate function in the base class.

File: rasa/utils/tensorflow/layers.py [hunk 4]
```diff
@@ -1085,3 +1140,314 @@ def call(
...
+
+        # Average the loss over the batch
+        return tf.reduce_mean(loss)
+
+    @staticmethod
+    def _accuracy(
+        sim_pos: tf.Tensor,  # (batch_size, 1, 1)
+        sim_candidates: tf.Tensor,  # (batch_size, 1, num_candidates)
+        pos_neg_labels: tf.Tensor,  # (batch_size, num_candidates)
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/layers.py:1433]: This one as well can be made to reuse between the two classes. For `SingleLabelDotProductLoss`, we can pass `sim_pos` as `pos_neg_labels`.

**Review by dakshvar22 — CHANGES_REQUESTED** [review]
Reviewed the code part. The new classes look good, but I think we can reuse more code across the two.

File: rasa/utils/tensorflow/layers_utils.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        n_max: Maximum index
+
+    Returns:
+        A uniformly distributed integer tensor of indices
+    """
+    return tf.random.uniform(shape=(batch_size, n), maxval=n_max, dtype=tf.int32)
+
+
+def batch_flatten(x: Tensor) -> Tensor:
+    """Makes a tensor 2D.
```
Comment (samsucik) [inline rasa/utils/tensorflow/layers_utils.py:21]: You could take what you wrote under `Returns` and move a shortened version of it here for better summary 🙂

File: rasa/utils/tensorflow/layers_utils.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+) -> Tensor:
+    """Creates a tensor of shape `target_shape` by padding it with `value` on the right.
+    
+    Args:
+        x: Any tensor
+        target_shape: Shape of the padded x; must be at least as large as the
+            shape of x in all dimensions
+
+    Returns:
+        A tensor like x, but padded with zeros
```
Comment (samsucik) [inline rasa/utils/tensorflow/layers_utils.py:44]:
```suggestion
        A tensor like x, but padded with the provided `value`.
```

File: rasa/utils/tensorflow/models.py [hunk 7]
```diff
@@ -577,7 +577,7 @@ def _prepare_ffnn_layer(
     def _prepare_dot_product_loss(
```
Comment (samsucik) [inline rasa/utils/tensorflow/models.py:577]: I'd argue that keeping a method with such a general name is a bit unfair now that it creates a single-label loss specifically. I know that we use single-label everywhere, but then one could ask why a model like IntentTED, which is so similar to TED, can't use a general method like this one, which TED uses... What do you think? Does it make sense to rename it to `_prepare_single_label_dot_product_loss` or to add an arg that determines which loss type will be used, so that this method can be used with both types?

File: tests/utils/tensorflow/test_layers_utils.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
+import pytest
```
Comment (samsucik) [inline tests/utils/tensorflow/test_layers_utils.py:1]: I think this import is unused.

File: tests/utils/tensorflow/test_layers_utils.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+import tensorflow as tf
+import numpy as np
+import rasa.utils.tensorflow.layers_utils as layers_utils
+
+
+def test_random_indices():
+    indices = layers_utils.random_indices(10, 4, 100)
+    assert np.all(tf.shape(indices).numpy() == [10, 4])
+    assert np.max(indices.numpy()) <= 100
```
[the remark this answers is not in the document]
Comment (samsucik) [reply tests/utils/tensorflow/test_layers_utils.py:10]: Also, you might want to add a check where `num_candidates=0`.

**Review by samsucik — CHANGES_REQUESTED** [review]
Awesome comments and I appreciate the shapes in the newly added code!

I left many comments that relate to code which you didn't actually change, but you definitely made active use of it by moving it under a different class. For those cases, I'm not sure if it's reasonable to ask you to make improvements to that code. Probably not, as it would be a huge rabbit hole. But I still left at least some suggestions for now in case you're up for it.

Especially clarifying the shapes would help a lot in chunks of the old or recycled code. Or sometimes it may not be necessary to always specify the input/output shapes of a method, but at least adding an explanatory comments whenever there's a `reshape`, `expand_dims`, etc. would be really helpful.

Btw I left some comments for the parent class and then didn't repeat them for the other two subclasses, but often they're really applicable to all the classes 🙂

File: rasa/utils/tensorflow/layers.py [hunk 10]
```diff
@@ -618,6 +619,120 @@ def f1_score(
...
+        constrain_similarities: bool = True,
+        model_confidence: Text = SOFTMAX,
+        similarity_type: Text = INNER,
+        name: Optional[Text] = None,
+    ):
+        """Declare instance variables with default values.
+
+        Args:
+            num_candidates: Positive integer, the number of incorrect or candidate labels;
+                the algorithm will minimize the similarity of negative labels to the input.
```
[the remark this answers is not in the document]
Comment (JEM-Mosig) [reply rasa/utils/tensorflow/layers.py]: > incorrect (or "candidate") labels 

That's actually not what I meant. Incorrect and candidate labels are not the same, unless its a single-label loss. I've clarified as follows now:
```
num_candidates: Number of labels besides the positive one. Depending on
                whether single- or multi-label loss is implemented (done in 
                sub-classes), these can be all negative example labels, or a mixture of
                negative and further positive labels, respectively.
```

File: rasa/utils/tensorflow/layers.py [hunk 11]
```diff
@@ -1085,3 +1140,314 @@ def call(
...
+            constrain_similarities: Boolean, if 'True' applies sigmoid on all
+                similarity terms and adds to the loss function to
+                ensure that similarity values are approximately bounded.
+                Used inside _loss_cross_entropy() only.
+            model_confidence: Model confidence to be returned during inference.
+                Possible values - 'softmax' and 'linear_norm'.
+
+        Raises:
+            LayerConfigException: When `similarity_type` is not one of 'cosine' or
+                'inner'.
```
[the remark this answers is not in the document]
Comment (JEM-Mosig) [reply rasa/utils/tensorflow/layers.py:1168]: No, I think `Raises` should be mentioned. At least they are recommended in the python code conventions on Notion.

File: rasa/utils/tensorflow/layers.py [hunk 12]
```diff
@@ -1085,3 +1140,314 @@ def call(
...
+            total_candidates=tf.shape(all_labels_embed)[0],
+        )
+
+        # Get the label embeddings and ids corresponding to candidate indices
+        candidate_labels_embed = self._get_candidate_values(
+            all_labels_embed, candidate_ids
+        )
+        candidate_labels_ids = self._get_candidate_values(all_labels_ids, candidate_ids)
+
+        # Determine how many distinct labels exist (highest label index)
```
[the remark this answers is not in the document]
Comment (JEM-Mosig) [reply rasa/utils/tensorflow/layers.py:1295]: The most general way would be to convert `all_labels_ids` into a python set and measure its lengh. But this would be slow and move outside the TF graph. I think the way it is done now is good. If we ever use strings or other objects as label IDs, much of the other code wouldn't work either.

File: tests/utils/tensorflow/test_layers_utils.py (hunk 8 above)
Comment (JEM-Mosig) [reply tests/utils/tensorflow/test_layers_utils.py:1]: It is used in line 47.

File: rasa/utils/tensorflow/layers.py [hunk 13]
```diff
@@ -670,39 +785,25 @@ def __init__(
...
-        )
 
     @staticmethod
-    def _sample_idxs(batch_size: tf.Tensor, x: tf.Tensor, idxs: tf.Tensor) -> tf.Tensor:
+    def _sample_idxs(
+        batch_size: tf.Tensor, x: tf.Tensor, idxs: tf.Tensor
+    ) -> tf.Tensor:  # ToDo: abstract this together with _get_candidate_values
         """Sample negative examples for given indices"""
 
         tiled = tf.tile(tf.expand_dims(x, 0), (batch_size, 1, 1))
```
[the remark this answers is not in the document]
Comment (JEM-Mosig) [reply rasa/utils/tensorflow/layers.py:822]: @samsucik I'd like to do this in a separate PR.

File: rasa/utils/tensorflow/models.py (hunk 7 above)
Comment (JEM-Mosig) [reply rasa/utils/tensorflow/models.py:577]: That will be handled in [Daksh's PR](https://github.com/RasaHQ/rasa/pull/8555/files#diff-1b0d53e90a7f1dcca1113cda22c58c98685004f8dfc7ba3d4e5bf9cee78e6a19R593)

File: rasa/utils/tensorflow/layers_utils.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+        n: Number of random indices in each example
+        n_max: Maximum index
+
+    Returns:
+        A uniformly distributed integer tensor of indices
+    """
+    return (
+        tf.random.uniform(shape=(batch_size, n), maxval=n_max, dtype=tf.int32)
+        if n_max > 0
+        else tf.zeros((batch_size, n), dtype=tf.int32)
```
[the remark this answers is not in the document]
Comment (JEM-Mosig) [reply rasa/utils/tensorflow/layers_utils.py:20]: Right. I added this because Sam asked me to add a test for when `n_max` is zero, and `tf.random.uniform` doesn't allow this.

[the remark this answers is not in the document]
Comment (dakshvar22) [reply rasa/utils/tensorflow/layers_utils.py]: if n_max <= 0, we should raise an exception IMO. It doesn't even make sense to return anything?

[the remark this answers is not in the document]
Comment (JEM-Mosig) [reply rasa/utils/tensorflow/layers_utils.py]: 🤦 the test is actually wrong. The max value is actually excluded and so it indeed doesn't make sense. Changed the test.

File: tests/utils/tensorflow/test_layers.py [hunk 15]
```diff
@@ -1,7 +1,245 @@
...
+    )
+    # The inputs just stay the inputs, up to an extra dimension
+    assert np.all(
+        pos_inputs_embed.numpy() == tf.expand_dims(batch_inputs_embed, axis=-2).numpy()
+    )
+    # The first example labels of each batch are in `pos_labels_embed`
+    assert np.all(pos_labels_embed.numpy() == np.array([[[l0]], [[l2]], [[l3]]]))
+    # The candidate label embeddings are picked according to the `mock_indices` above.
+    # E.g. a 2 coming from `mock_indices` means that the first positive label (always)
+    # of example 2 (`[l3, l0, _]`) is picked, i.e. `l3`.
```
Comment (samsucik) [inline tests/utils/tensorflow/test_layers.py:207]: Same comment as above 😉

**Review by samsucik — APPROVED** [review]
Looks good to me (I looked especially at the tests).

One small thing I noticed is that the loss classes are often initialised by passing to their constructors a lot of args, somw of which could be left out to simplify the tests.

Another thing is that those 2 test cases are reaaaally long, but they also flow quite nicely (also considering the in-line comments), so I wouldn't insist on splitting them up into smaller bits, only if you've got some good ideas for how to do it 🙂

File: rasa/utils/tensorflow/layers.py [hunk 16]
```diff
@@ -1098,23 +1098,26 @@ def _chosen_loss(self) -> Callable:
...
-            all_labels_embed: Embedding tensor for the all labels.
-            all_labels: Tensor representing all labels.
+            inputs_embed: Embedding tensor for the batch inputs;
+                shape (batch_size, ..., num_features)
+            labels_embed: Embedding tensor for the batch labels;
+                shape (batch_size, ..., num_features)
+            labels: Tensor representing batch labels; shape (batch_size, ..., 1)
+            all_labels_embed: Embedding tensor for the all labels;
+                shape (num_labels, num_features)
+            all_labels: Tensor representing all labels; shape (num_labels, 1)
```
[the remark this answers is not in the document]
Comment (JEM-Mosig) [reply rasa/utils/tensorflow/layers.py:1118]: I keep forgetting this!

File: tests/utils/tensorflow/test_layers.py [hunk 17]
```diff
@@ -1,7 +1,245 @@
...
+    )
+    # The inputs just stay the inputs, up to an extra dimension
+    assert np.all(
+        pos_inputs_embed.numpy() == tf.expand_dims(batch_inputs_embed, axis=-2).numpy()
+    )
+    # The first example labels of each batch are in `pos_labels_embed`
+    assert np.all(pos_labels_embed.numpy() == np.array([[[l0]], [[l2]], [[l3]]]))
+    # The candidate label embeddings are picked according to the `mock_indices` above.
+    # E.g. a 2 coming from `mock_indices` means that the first positive label (always)
+    # of example 2 (`[l3, l0]`) is picked, i.e. `l3`.
```
[the remark this answers is not in the document]
Comment (JEM-Mosig) [reply tests/utils/tensorflow/test_layers.py:115]: Right! It was different before and I changed it to pick things from all labels, but then forgot to update the comment. Good catch!

File: rasa/utils/tensorflow/layers.py [hunk 18]
```diff
@@ -1178,29 +1196,36 @@ def __init__(
...
-            all_labels_embed: Embeddings for all labels in the domain
-            all_label_ids: Indices for all labels in the domain
+                for IntentTED);
+                shape (batch_size, max_num_labels_per_input, num_features)
+            batch_label_ids: Batch label indices (e.g. indices of the intents). We
+                assume that indices are integers that run from 0 to
+                (number of labels) - 1.
+                shape (batch_size, max_num_labels_per_input, 1)
+            all_labels_embed: Embeddings for all labels in the domain;
+                shape (batch_size, max_num_labels_per_input, 1)
```
[the remark this answers is not in the document]
Comment (JEM-Mosig) [reply rasa/utils/tensorflow/layers.py:1219]: Right. Copy-paste error.

**Review by dakshvar22 — APPROVED** [review]
LGTM 💯  Can you please run model regression tests (both NLU and Core) before the final merge?

**Comment by JEM-Mosig** [comment]
> LGTM 💯 Can you please run model regression tests (both NLU and Core. You might want to merge `intent-ted` into your branch before running them. I just merged `main` into `intent-ted`.) before the final merge?

On it. I'll merge `intent-ted` into here and start all regression tests.

**Comment by JEM-Mosig** [comment]
/modeltest
```yml
include: 
  - dataset: ["all-nlu"]
    config: ["all-nlu"]
  - dataset: ["all-core"]
    config: ["Rules + TED", "Rules + AugMemo + TED", "Rules + Memo + TED"]
```
